Server UI Changes (#677)
jamulussoftware/jamulussoftware.github.io: wiki/en/Running-a-Server.md

The short answer is **no**. For various reasons, you should try to use other people's servers, at least at first while getting your sound set up with the Jamulus client.

> Note that you can have a "private" session with other people on a Server by simply soloing each other. You will then not be able to hear anyone else if they enter your server.

Reviewer: Is it worth adding that this doesn't affect things like chat and profile information being seen by others?  Or is that too obvious?

Author: I think the understanding of "mute" should be enough, but we could add something to clarify.

- Running a server may require you to adjust any firewalls running on or outside of your machine or cloud host.

> - Running an Uregistered Server at home will require you to [port forward](#running-an-unregistered-server) on your router. When running a Registered Server, port forwarding should not be necessary in most cases, but it's advisable to do so because some networks may not work properly with Jamulus in its default mode.

Reviewer: Ah, I see the port forwarding "problem" is mentioned here already :) .

Author: Yes, a relatively recent addition. These docs haven't been changed much otherwise in almost two years I think.

- Jamulus only has limited IPv6 support which needs to be enabled with a command line argument on the client and server. There are plans to expand IPv6 support.

# Server Types

You can run your server in different ways (either at home or on a 3rd party host):

### 1. Registered 

> Your server will be listed in a Directory. By default, Jamulus has a list of Public Directories that clients can connect to. If you register with one of these, anyone can then discover and connect to your server. 

Reviewer: Check for trailing whitespace?

> Your server will be listed in a Directory. By default, Jamulus has a list of Public Directories that clients can connect to. If you register with one of these, anyone can then discover and connect to your server. 

Reviewer 2: ```suggestion
Your server will be listed in a Directory. By default, Jamulus has a list of Public Directories that clients can connect to. If you register with one of these, your server is believed to be public and anyone can then discover and connect to your server. 
```

Not happy with my change, but we should stress the "public" part here.

Author: This might sort of clash with the FAQ entry on this, which says that some servers on the public Directories are sort of private.

Reviewer 2: > In general, if somebody sets up a public server, they accept that anyone can play on it. Jamulus has no password protection or other authentication mechanisms. However, some servers may state their policies in the welcome message you will see in the chat window.

Implies that mostly all servers registered at a public directory are believed to be freely available

Reviewer: > your server is believed to be public and

You can set up a firewall rule that lets you register your Server with Any Genre 1 but (unless `--showallservers` is used) won't be listed in the Connect window.  So adding any extras like this only confuses matters.

I agree _in principle_ that anyone using the Public Directories _should_ be registering their server for public use -- but we don't require it.

<figure>
	<img src="{% include img/en-screenshots/diagram-reg-server.inc %}" loading="lazy" alt="Diagram of connections between clients within a Jamulus Registered Server">
<figcaption>How Registered Servers work</figcaption>
</figure>

### 2. Unregistered 

> This is the default when starting a server for the first time. Unregistered Servers are not listed by Directories, so only musicians who know your server's address to will be able to connect to it. This is useful because Jamulus does not let you control who can connect to a server.

Reviewer: > This is the default when starting a server for the first time.

Not yet :).  By default - if running with the UI - you're registered with Any Genre 1.

If you run headless, you'll not register unless you supply a directory address (because headless servers don't have inifiles, not supplying a directory is the only way to say "don't register me with Any Genre 1").

Reviewer: (I can't remember if I kept this existing behaviour -- I think I did...  It's probably worth including the change if I didn't already...)

Author: Do we not intend to make it so that the default with the GUI is unregistered?

Reviewer: Yep, but my PR as it stands hasn't changed legacy behaviour.  I'd also need to review the messing about when the settings are loaded to make sure everything still worked as intended.

I can do it in https://github.com/jamulussoftware/jamulus/pull/2199 ... probably won't be today.

Reviewer: Jamulus 3.8.1 GUI server on Windows registers by default with Any Genre 1.

And... another new profile...
![image](https://user-images.githubusercontent.com/1549463/148696477-49567b99-aaa6-4f11-ab59-d5c6b4a52081.png) 
![image](https://user-images.githubusercontent.com/1549463/148696490-632e9c87-c99f-4333-bbef-b14c8ad95667.png)

**However**... now I check, Jamulus _master_ (ff281ca1 2022-01-08) on Linux doesn't...

So I take this all back, I must have done it in one of the "OCD re-order" changes as a "freebie add-on".  Or accident.

> This is the default when starting a server for the first time. Unregistered Servers are not listed by Directories, so only musicians who know your server's address to will be able to connect to it. This is useful because Jamulus does not let you control who can connect to a server.

Reviewer 2: ```suggestion
This is the default when starting a server for the first time. Unregistered Servers are not listed by Directories, so only musicians who know your server's address to will be able to connect to it. This is useful if you don't want your server to be public since Jamulus does not let you control who can connect to a server.
```

... or published

Author: Again, the use of the term "public" there might be misleading because you might be connecting your server to a private directory (eg on a LAN). This kind of "correctness" is the point of this PR.

Reviewer: > if you don't want your server to be public since

I think the simpler version ("because") is clearer and less confusing (without any change).

<figure>
	<img src="{% include img/en-screenshots/diagram-unreg-server.inc %}" loading="lazy" alt="Diagram of connections between clients within a Jamulus Unregistered Server">
	<figcaption>How Unregistered Servers work</figcaption>
</figure>

### 3. Custom Directory

If you want to run a number of Servers, possibly also behind a firewall or on a LAN, you may want to run your server as a Directory. Examples include online events, music associations, sectional rehearsals or music lessons for schools.

> To view servers listed by a Custom Directory, musicians must enter its address in their client’s "Custom Directory Server" settings field.  They will then see a Connection Setup list which is generated by your Directory. Custom Directories otherwise work for clients in the same way as Public Directories.

Reviewer: > client’s "Custom Directory Server" settings field

![image](https://user-images.githubusercontent.com/1549463/148685228-f12728ac-0d94-49ac-b380-25baa4c11823.png)
"Custom Directories", as we store multiple at the client.

Reviewer: > They will then see a Connection Setup list

![image](https://user-images.githubusercontent.com/1549463/148685326-02c68912-39d2-4279-a16f-7e96ee044836.png)
It's just the "Connect" window, where the server list is supplied by the chosen Directory.

Reviewer: On Linux master, I get
![image](https://user-images.githubusercontent.com/1549463/148696164-5bdf2832-6c10-4030-98fd-21c2f2849ab1.png)

I'm not sure where the (local) hostname is coming from.

Author: Sorry, does your comment imply a change to the wording here or not?

Reviewer: Yes.  The text is wrong as it stands.

Author: OK I've re-written this in the new Custom-Directories.md file.

Up to 150 servers can then register with your Directory by setting your Directory address in their `--directoryserver` option at startup.

> To run a server as a Directory, it should be configured with `--directoryserver` (that is, making itself the Directory to query for servers). When running a server with the GUI, select "Custom" from the Directory drop-down list. 

Reviewer 2: @pljones is this true? Don't we need --directoryserver localhost or -e localhost anymore?

Reviewer: No.  I've raised it in many places.

> To run a server as a Directory, it should be configured with `--directoryserver localhost` (that is, making itself the Directory to query for servers). When running a server with the GUI, select "Custom" from the Directory drop-down list. 

Reviewer: > select "Custom" from the Directory drop-down list

... and enter "localhost" as the Custom Directory server address.

Author: Is there a reason why it doesn't do that automatically?

Reviewer: The _vast_ majority of people registering with a Custom Directory are doing so as Servers.

They're _not_ setting up a Directory.

Otherwise you'd end up with more Directories than Servers... which makes no sense at all.

Author: Sorry, I meant when they choose the "Directory" option in the Directory drop-down. Or have I misunderstood what that means? So right now if I want to set my server up as a Directory, I have to declare that in two places in the UI?

Reviewer: There is no "Directory" option in the Server "Directory" drop down.  You pick "Custom" and enter one of the two literal strings in the Custom Directory server address that you'd pass to the `-e` option on the command line.

Basically, "Custom" works the same way as the command line option.  ("None" works like not giving the `-e` on the command line; the genre values aren't available from the command line.)

Reviewer: It might be simpler to strip _all_ mention of running as a Directory from the documentation except for the definition list and from there have a link to a single page that explains how to set one up both from the command line (the normal thing you'd do) and from the GUI.

Almost no one wants to know.

Reviewer 2: Yes. I think so too. Probably we'd also stress the unregistered server part in the docs.
Setting up a registered (public) server is probably no longer that useful. We have a bunch of free ones all over the world (which are probably all running via CLI by more experienced users).

Author: Good idea. It's too confusing trying to keep it all together as different "states" of the server.

> #### Points to note about Custom Directories

Reviewer: > Custom

Doesn't make a difference.  Several of the Public Directories use `--listfilter` (using the undocument version filter).

- If you want to control which servers can register with your Directory, you can enable a whitelist with the `--listfilter` command line option. See the section on command line options later in this document.

> - When running a Directory behind a NAT firewall on a private network, use the `--serverpublicip` option to specify the public IP address of the server(s) being listed by your Directory. This is necessary to allow clients on the public Internet to connect to them via NAT. Note that for the servers using this option, you will still need proper port forwarding in your router/firewall.

Reviewer: >  public IP address of the server(s) being listed by your Directory

Only applies to servers on the same LAN / with the same public IP.  The phrasing implies it applies to _all_ registered servers.

<figure><img src="{% include img/en-screenshots/server-window-setup.inc %}" style="width:80%; border:5px solid grey;" loading="lazy" alt="Image of the Jamulus server setup window"></figure>

### The Directory list 

> **None**: By default, you will not be connected to a directory and will be running in unregistered mode. [Read these instructions](#running-an-unregistered-server) to have other people connect to your server in this mode.

Reviewer 2: ```suggestion
**None**: By default, you will not be connected to a directory and will be running in unregistered mode. This means your server will not be publically visible. [Read these instructions](#running-an-unregistered-server) to have other people connect to your server in this mode.
```

Author: I think attempts to resurrect the ideas of public and private may not be right. But let's see. You can run an unregistered server and make it public by putting its address out in some other way, for instance.

Reviewer 2: You can, but then you know that you're publishing it. I think this discussion should go to https://github.com/jamulussoftware/jamulus/discussions/2228

Reviewer: >  This means your server will not be publically visible.

I wouldn't bother with the addition.  Simply registering doesn't mean others will see you.  You'd get into the discussion of different types of Directory here, which isn't wanted.

> **None**: By default, you will not be connected to a directory and will be running in unregistered mode.  [Read these instructions](#running-an-unregistered-server) to have other people connect to your server in this mode.

Reviewer: See previous note - currently the GUI registers with Any Genre 1.

Reviewer: I wuz wrong.

**Genre**: To allow other people to see your server on one of the built-in public directories, select your desired genre directory. You should see a confirmation message saying whether your server has registered successfully. If not, and you leave your server running, it will keep trying to register until a free slot becomes available.  

> **Custom**: This sets your server to be a directory on which other servers can be listed. Note that when setting this mode, you may want also to specify the location of server list file (See "Options" below) unless you want to disconnect all listed servers when restarting Jamulus.

Reviewer: > This sets your server to be a directory on which other servers can be listed.

As noted else where, this is wrong.

Custom Directories are where Servers register so they can be seen by Clients viewing that Custom Directory in the Connect window.

Setting the select box to "Custom" and setting a Custom Directory server address is how to do this.

For the tiny number of operators running Directories, putting "localhost" into the Custom Directory server address means you're running as a Custom Directory yourself, rather than registering your server with a Custom Directory.

> **Custom**: This sets your server to be a directory on which other servers can be listed. Note that when setting this mode, you may want also to specify the location of server list file (See "Options" below) unless you want to disconnect all listed servers when restarting Jamulus.  

Reviewer: > **Custom**: This sets your server to be a directory on which other servers can be listed. 

Wrong.

Custom allows you to set and use a Custom Directory server address.  Nothing more than that.

"Registering" with localhost is what makes a server a directory.

Author: See https://github.com/jamulussoftware/jamuluswebsite/pull/677#discussion_r780814854 This seems a rather odd UX.

Reviewer: Why?

First off, most _servers_ are run headless, as far as I know.

No one wants to run as a Directory apart from a handful of people.

1. Most people running a server register with a Public Directory (Jamulus built in address) - they pick a Genre.

2. Those not wanting to register at all pick None.

3. The next most common group is those who are registering with a Custom Directory - they select Custom here (and put the Custom Directory server address in the "Custom Directory server address" field).

4. The least common group is those who are running Directories (both the Public and Custom ones -- same deal).  Of these, a tiny fraction will want to do it using the GUI (so a tiny fraction of a tiny fraction).
Just like from the command line, they use a Custom Directory server address of "localhost" or "127.0.0.1" -- because that's how we identify running as a Directory -- and, just like any other GUI server operator they'll select Custom for Directory and enter the Custom Directory server address.

There's _no point_ making running as a GUI Directory a special case as it's so rare.

Author: My point was simply one of basic usability, but if we take the view that it's rare enough not to matter (and we're putting the Directory docs into another place now anyway) then that's OK.

### My Server Info

> When running as a registered server this displays the server's name, city and country so that other users can easily identify it in the directory listing.

Reviewer: > this displays

Not really the right way to say it.  This is the server set up screen.  It's not "displaying" it - it's allowing it to be set by the server operator.

"Display" says "look but don't touch".

Author: The fact they can be edited in the UI here is self-evident (it's a form) and is consistent with how we talk about My Profile profile too.

Reviewer: But it's still wrong.  Is it okay to use bad English just because you use it all the time?

Author: It's standard practice to explain the _effect_ of using a UI, not _how_ you use it (unless your UI is unusable). So if you say "You can set the server's name, city and country... " then that begs the question why and how, which is the more important thing to explain.

**Note**: When your server is recording, clients will display a message that recording is on.

### Custom Directory Server Address

Leave this field empty unless you need to list your server on a [Custom Directory](#3-directory)

### Server List Filename

> If you are running your server as a custom directory and need to restart it for any reason (for example when rebooting the host), servers connected to it will be disconnected until they re-register. This does not mean that clients connected to those servers will be disconnected, but does mean that new clients will be not able to see servers listed by your directory until those servers reconnect. To enable the list of registered servers to persist between restarts, specify the location and name of a file that the directory can read and write to here.  

Reviewer: > as a custom directory

Just "Directory".  Doesn't need to be a Custom one.

### Delay panning

> This option uses small differences in sound arrival time between the two ears. It produces a stereo effect similar to natural human hearing when compared to normal "volume" panning. Note that to achieve the effect, all clients should be set to 'mono in / stereo out' (or to 'stereo', but only if transmitting a stereo signal).

Reviewer: > Note that to achieve the effect, all clients should be set to 'mono in / stereo out' (or to 'stereo', but only if transmitting a stereo signal).

I think that's redundant.  To hear a stereo effect, you need stereo out.  Goes without saying?  (Yes, cut-n-paste from above, I can see that.)

It also implies that the effect doesn't work for any of the clients if one of the clients has mono out.  I don't think that's even true as the mixer runs independently for each client output.

Author: I recall somebody was very keen to add this line when we originally did it (can't remember who) but if it's not accurate we should correct.

### Start Minimised

**Windows users** - If you want the server to start automatically on system start, enable the corresponding check box.

# Running an Unregistered Server

> It is highly recommended to test your server on a **Public Directory** first so as to narrow down any subsequent problems in uregistered mode.

Reviewer 2: ```suggestion
It is highly recommended to test your server on a **Public Directory** first so as to narrow down any subsequent problems in unregistered mode.
```

### Port forwarding 

> If you are running a server on your home network, people from outside will not be able to see things inside it. So if you want to operate a Jamulus server, you need to open a hole in your router to let Jamulus clients connect to it   

Reviewer: More trailing whitespace - and missing punctuation at end of sentence.
